certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jdbc/UserExtractor.java

```diff
@@ -0,0 +1,110 @@
+
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Extract {@link List<User>} from the {@link ResultSet}.
+ */
+@Component
+public class UserExtractor implements ResultSetExtractor<List<User>> {
+    @Autowired
+    private CertificateInJsonMapper certificateInJsonMapper;
+
+    public void setCertificateInJsonMapper(CertificateInJsonMapper certificateInJsonMapper) {
+        this.certificateInJsonMapper = certificateInJsonMapper;
+    }
+
+    /**
+     * Extract {@link List<User>} from the {@link ResultSet}.
+     *
+     * @param resultSet is the {@link ResultSet} to map.
+     * @return {@link List<User>}.
+     * @throws SQLException        when something went wrong.
+     * @throws DataAccessException when the datasource is not available.
+     */
+    @Override
+    public List<User> extractData(ResultSet resultSet) throws SQLException, DataAccessException {
+        List<User> users = new ArrayList<>();
+
+        while (resultSet.next()) {
+            final long currentId = resultSet.getLong(ColumnNames.TABLE_USER_COLUMN_ID);
+            final long currentOrderId = resultSet.getLong(ColumnNames.TABLE_USERORDER_COLUMN_ID);
+            if (users.stream().anyMatch(user -> user.getId() == currentId)) {
+                User user = users.stream().filter(user1 -> user1.getId() == currentId).findAny().get();
+                if (user.getOrders().stream().anyMatch(order -> order.getId() == currentOrderId)) {
+                    addGiftCertificateToCorrespondingOrderOfUserOrders(resultSet, users, currentId, currentOrderId);
+                } else {
+                    Order newOrder = createNewOrderFromResultSetLine(resultSet);
+                    if (resultSet.getString(ColumnNames.TABLE_USERORDER_CERTIFICATE_COLUMN_CERTIFICATEINJSON) != null) {
+                        GiftCertificate giftCertificate = certificateInJsonMapper.mapRow(resultSet, resultSet.getRow());
+                        newOrder.getCertificates().add(giftCertificate);
+                        users.stream().filter(user2 -> user2.getId() == currentId)
+                                .findAny().ifPresent(user2 -> user2.getOrders().add(newOrder));
+                    }
+                }
+            } else {
+                List<Order> allOrdersThisUser = new ArrayList<Order>();
+                User user = createNewUserFromResultSetLine(resultSet, allOrdersThisUser);
+                if (resultSet.getLong(ColumnNames.TABLE_USERORDER_COLUMN_ID) != 0) {
+                    Order newOrder = createNewOrderFromResultSetLine(resultSet);
+                    user.getOrders().add(newOrder);
+                    users.add(user);
+                    addGiftCertificateToNewOrderOfUsersOrders(resultSet, users, currentId, currentOrderId, newOrder);
+                } else {
+                    users.add(user);
+                }
+            }
+        }
+        return users;
```

Comment:
Divide this method into many private methods because there are sooo many if statements!

Reply:
done